### PR TITLE
NetSim: Support Peer-to-Peer Connections

### PR DESCRIPTION
**Pull request by islemaster**
Up until now, NetSim has been about connecting to a router.  This PR makes it possible to connect directly to another client node, and send messages back and forth.

out

Because the lobby was the first part of NetSim I wrote, it was a little painful to work with and I did a lot of refactoring.  Big positive changes there:
- Actual reuse of `NetSimPanel`
- The whole lobby interface is now safe to re-render any time.
- Separation and cleanup of shard selection from remote node selection.

Another major change here is that I've removed `NetSimConnection` entirely.  Its role was always a little unclear.  The core `NetSim` class inherited most of its functionality, detecting changes to connection status and triggering related re-rendering.

There's a new levelbuilder control here: You can make a node type (routers, clients) visible in the lobby but not connectable.  Not super useful, but included for completeness.

### CODE REVIEW

File: apps/src/netsim/NetSimLobby.js [hunk 1]
```diff
@@ -48,415 +61,324 @@ var NetSimLobby = module.exports = function (levelConfig, connection, user,
...
-NetSimLobby.prototype.onRowClick_ = function (listItem, connectionTarget) {
-  // Can't select user rows (for now)
-  if (NodeType.CLIENT === connectionTarget.getNodeType()) {
-    return;
-  }
+NetSimLobby.prototype.onWireTableChange_ = function (rows) {
+  // Update the collection of nodes with connections pointing toward us.
+  this.nodesRequestingnetsim_ = rows.filter(function (wireRow) {
+    return wireRow.remoteNodeID === this.myNode_.entityID;
+  }.bind(this)).map(function (wireRow) {
```
Comment (bcjordan) [inline apps/src/netsim/NetSimLobby.js]: I could see either way, just curious—any reason / preference for using `.bind(this)` over using `this` as the second param to `filter`, like `.filter(function(){}, this)`?

Comment (bcjordan) [inline apps/src/netsim/NetSimLobby.js]: Around here since there are also `_` functions it does provide a sort of consistency and clarity / harder-to-mess-up-ness.

File: apps/src/netsim/NetSimLobby.js [hunk 2]
```diff
@@ -48,415 +61,324 @@ var NetSimLobby = module.exports = function (levelConfig, connection, user,
...
  * @private
  */
-NetSimLobby.prototype.onRowClick_ = function (listItem, connectionTarget) {
-  // Can't select user rows (for now)
-  if (NodeType.CLIENT === connectionTarget.getNodeType()) {
-    return;
-  }
+NetSimLobby.prototype.onWireTableChange_ = function (rows) {
+  // Update the collection of nodes with connections pointing toward us.
+  this.nodesRequestingnetsim_ = rows.filter(function (wireRow) {
```
Comment (bcjordan) [inline apps/src/netsim/NetSimLobby.js]: camel case as s/nodesRequestingnetsim_/nodesRequestingNetsim_ ?

File: apps/src/netsim/NetSimPanel.html.ejs [hunk 3]
```diff
@@ -2,7 +2,9 @@
      class="netsim-panel <%= className %>">
   <h1>
     <span class="minimizer">
+      <% if (canMinimize) { %>
       <i class="fa fa-minus-square"></i>
```
Comment (bcjordan) [inline apps/src/netsim/NetSimPanel.html.ejs]: should indent? (not sure)

File: apps/src/netsim/NetSimRemoteNodeSelectionPanel.html.ejs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+ * @property {string} displayName
+ * @property {string} status
+ */
+
+/**
+ * @param {NetSimNode} node
+ * @returns {boolean}
+ * @private
+ */
+var isSelectedNode = function (node) {
```
Comment (bcjordan) [inline apps/src/netsim/NetSimRemoteNodeSelectionPanel.html.ejs]: Should/could this stuff be pulled up out of the `.html.ejs` template into the controller?

File: apps/src/netsim/NetSimRemoteNodeSelectionPanel.js [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+var utils = require('../utils');
+var _ = utils.getLodash();
+var i18n = require('../../locale/current/netsim');
+var NetSimPanel = require('./NetSimPanel');
+var markup = require('./NetSimRemoteNodeSelectionPanel.html');
+var NodeType = require('./netsimConstants').NodeType;
+
+/**
+ * Generator and controller for lobby/connection controls.
```
Comment (bcjordan) [inline apps/src/netsim/NetSimRemoteNodeSelectionPanel.js]: Comment need updating?

File: apps/src/netsim/NetSimRemoteNodeSelectionPanel.js [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+ * @param {number} myNodeID
+ * @param {function} addRouterCallback
+ * @param {function} selectNodeCallback
+ * @param {function} connectButtonCallback
+ * @param {function} cancelButtonCallback
+ * @constructor
+ * @augments NetSimPanel
+ */
+var NetSimRemoteNodeSelectionPanel = module.exports = function (rootDiv,
+    levelConfig, nodesOnShard, nodesRequestingConnection, selectedNode,
```
Comment (bcjordan) [inline apps/src/netsim/NetSimRemoteNodeSelectionPanel.js]: Potentially worth pulling some of these into a config object?

File: apps/src/netsim/NetSimRemoteNodeSelectionPanel.js [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+    return false;
+  }
+
+  // Permissible connection limited by level configuration
+  return (
+      this.levelConfig_.canConnectToClients &&
+      connectionTarget.getNodeType() === NodeType.CLIENT
+      ) || (
+      this.levelConfig_.canConnectToRouters &&
+      connectionTarget.getNodeType() === NodeType.ROUTER );
```
Comment (bcjordan) [inline apps/src/netsim/NetSimRemoteNodeSelectionPanel.js]: Indentation here seems a little hard to parse, could throw this `);` on a newline, & maybe de-indent the `) || (` one? Alternatively could extract variable to name one or more of the inner boolean statements

File: apps/src/netsim/NetSimShardSelectionPanel.html.ejs [hunk 8]
```diff
@@ -0,0 +1,37 @@
+<%
+  var i18n = require('../../locale/current/netsim');
+%>
+<div class="content-wrap">
+  <div class="field-box display-name-control">
+    <label for="netsim_lobby_name"><%= i18n.myName() %></label>
+    <input id="netsim_lobby_name" type="text" value="<%= displayName %>" <% if (displayName.length > 0) { %>disabled<% } %> />
```
Comment (bcjordan) [inline apps/src/netsim/NetSimShardSelectionPanel.html.ejs]: Nit: usually prefer `-` over `_` for IDs/classes

**Comment by davidsbailey** [comment]
In the ani-gif, it would seem more intuitive for the left screen to visually show the connection being established only once the right screen has accepted the invitation, rather than showing it immediately after the left screen extends the invitation.

File: apps/src/netsim/NetSimLobby.js (hunk 1 above)
Comment (islemaster) [inline apps/src/netsim/NetSimLobby.js]: I always forget that these methods take a `thisArg`; it seems anti-idiomatic when I have to `.bind(this)` everywhere else.  That said, I've been trying to adopt this when using Array.prototype.forEach, so I should probably do so here as well.

File: apps/src/netsim/NetSimLobby.js (hunk 2 above)
Comment (islemaster) [inline apps/src/netsim/NetSimLobby.js]: Oh man - that was a bad replace.  Thanks!

File: apps/test/netsim/NetSimConnection.js [hunk 9]
```diff
@@ -1,87 +0,0 @@
...
-var netsimTestUtils = require('../util/netsimTestUtils');
-var fakeShard = netsimTestUtils.fakeShard;
-var assertTableSize = netsimTestUtils.assertTableSize;
-
-var NetSimLogger = testUtils.requireWithGlobalsCheckBuildFolder('netsim/NetSimLogger');
-var NetSimConnection = testUtils.requireWithGlobalsCheckBuildFolder('netsim/NetSimConnection');
-var NetSimClientNode = testUtils.requireWithGlobalsCheckBuildFolder('netsim/NetSimClientNode');
-var NetSimRouterNode = testUtils.requireWithGlobalsCheckBuildFolder('netsim/NetSimRouterNode');
-
-describe("NetSimConnection", function () {
```
Comment (bcjordan) [inline apps/test/netsim/NetSimConnection.js]: Do these tests no longer apply?

File: apps/src/netsim/NetSimPanel.html.ejs (hunk 3 above)
Comment (islemaster) [inline apps/src/netsim/NetSimPanel.html.ejs]: My convention has been yes; not sure from our style guide what's correct for ejs though.

**Comment by bcjordan** [comment]
LGTM overall! Commented a handful of nits, feel free to take or leave any of them

File: apps/src/netsim/NetSimRemoteNodeSelectionPanel.html.ejs (hunk 4 above)
Comment (islemaster) [inline apps/src/netsim/NetSimRemoteNodeSelectionPanel.html.ejs]: Sure; putting this together, I was just tempted to pass the controller instance into the ejs template so that it could call utility methods like this.  Is that considered bad form?

Comment (bcjordan) [inline apps/src/netsim/NetSimRemoteNodeSelectionPanel.html.ejs]: That pattern seems reasonable, something like to Rails' `_helper.rb` view helpers. Could alternatively pass those utility methods in as individual properties to be explicit about what's provided to the view

File: apps/test/netsim/NetSimConnection.js (hunk 9 above)
Comment (islemaster) [inline apps/test/netsim/NetSimConnection.js]: No, they don't; the only method being tested here was the `getAllNodes` method, which was redundant with `netsimNodeFactory.nodesFromRows` so I removed it, and since I subsequently removed the NetSimConnection type altogether I dropped this test file.

File: apps/src/netsim/NetSimShardSelectionPanel.html.ejs (hunk 8 above)
Comment (islemaster) [inline apps/src/netsim/NetSimShardSelectionPanel.html.ejs]: I was familiar with this convention for classes (and am still slowly correcting throughout netsim), but not for IDs (I've been sticking to snake_case for those).  Do you have a style guide reference for this?

Comment (bcjordan) [inline apps/src/netsim/NetSimShardSelectionPanel.html.ejs]: Looks like in our doc we recommend https://github.com/thoughtbot/guides/tree/master/style/sass which suggests not using IDs for styling at least (no mention of JS selection, though).

In any case, I'm pretty sure across most of CSS-land dashes are preferred. The [Google style guide](http://google-styleguide.googlecode.com/svn/trunk/htmlcssguide.xml) suggests dashes for IDs

Comment (islemaster) [inline apps/src/netsim/NetSimShardSelectionPanel.html.ejs]: Aha: [This rule right here](http://google-styleguide.googlecode.com/svn/trunk/htmlcssguide.xml?showone=ID_and_Class_Name_Delimiters#ID_and_Class_Name_Delimiters)

I'll add a mention in our Code.org style guide under HTML/HAML

**Comment by islemaster** [comment]
A thing I was supposed to remember _before_ opening this PR: This change is likely to break the netsim-lobby UI test.  I'll fix that.

**Comment by islemaster** [comment]
@davidsbailey you have a good point about the visualization.  I'm trying to get that cleaned up now.

**Comment by islemaster** [comment]
Okay, made the requested visualization change: The remote node isn't pulled into the foreground until a mutual connection is established.  Here's an updated screencast:

out

File: apps/src/netsim/NetSimLobby.js [hunk 10]
```diff
@@ -48,415 +61,330 @@ var NetSimLobby = module.exports = function (levelConfig, connection, user,
...
-    return;
-  }
-
-  this.getUserSections_(function (data) {
-    if (0 === data.length) {
-      this.useRandomShard();
-      return;
+NetSimLobby.prototype.fetchInitialLobbyData_ = function () {
+  this.shard_.nodeTable.readAll(function (err, rows) {
+    if (!err) {
```
Comment (davidsbailey) [inline apps/src/netsim/NetSimLobby.js]: consider logging errors, or explaining why it's ok not to

**Comment by davidsbailey** [comment]
LGTM overall. nice work!
